File: tests/fixtures/http_server.py
"""Loopback HTTP/1.1 server used by the native HTTP provider tests."""

import socket
import sys
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer


class Handler(BaseHTTPRequestHandler):
    protocol_version = "HTTP/1.1"

    def do_GET(self):
        if self.path == "/redirect":
            self.send_response(302)
            self.send_header(
                "Location",
                f"http://localhost:{self.server.server_port}/authorization",
            )
            self.send_header("Content-Length", "0")
            self.end_headers()
            return
        if self.path == "/authorization":
            body = self.headers.get("authorization", "none").encode("ascii")
        elif self.path == "/large":
            body = b"0123456789abcdef" * (4 * 1024 * 1024 // 16)
        else:
            body = b"small response\n"
        self.send_response(200)
        self.send_header("Content-Type", "application/octet-stream")
        self.send_header("Content-Length", str(len(body)))
        self.send_header("X-Repeated", "one")
        self.send_header("X-Repeated", "two")
        self.end_headers()
        self.wfile.write(body)

    def do_POST(self):
        if self.path == "/early":
            self.send_response(413)
            self.send_header("Content-Length", "0")
            self.send_header("Connection", "close")
            self.end_headers()
            self.wfile.flush()
            # The point of this route is to answer without consuming the upload, and
            # the client is expected to stop sending once it sees the answer. What it
            # must not do is close on top of a receive buffer that still has bytes in
            # it: the kernel answers that with RST rather than FIN, and a reset
            # discards the response the client had already parsed. Half-close the
            # write side to signal the end, then read until the client's own close,
            # bounded so a client that ignores the answer cannot hang the server.
            try:
                self.connection.shutdown(socket.SHUT_WR)
                self.connection.settimeout(5)
                while self.rfile.read(65536):
                    pass
            except OSError:
                pass
            self.close_connection = True
            return
        length = int(self.headers.get("content-length", "0"))
        body = self.rfile.read(length)
        self.send_response(200)
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def log_message(self, *_args):
        pass


server = ThreadingHTTPServer(("127.0.0.1", 0), Handler)
with open(sys.argv[1], "w", encoding="ascii") as port_file:
    port_file.write(str(server.server_address[1]))
server.serve_forever()

File: tests/fixtures/http_server.mjs
// Loopback HTTP/1.1 server used by the native HTTP provider tests.

import { writeFileSync } from "node:fs";
import { createServer } from "node:http";

const portFile = process.argv[2];
if (!portFile) {
  console.error("usage: node http_server.mjs PORT_FILE");
  process.exit(2);
}

const largeBody = Buffer.from("0123456789abcdef".repeat(4 * 1024 * 1024 / 16));
let loopHits = 0;

function send(res, status, body = Buffer.alloc(0), headers = {}) {
  res.writeHead(status, {
    "Content-Length": body.length,
    ...headers,
  });
  res.end(body);
}

const server = createServer((req, res) => {
  if (req.method === "GET") {
    if (req.url === "/redirect") {
      const { port } = server.address();
      send(res, 302, undefined, {
        Location: `http://localhost:${port}/authorization`,
      });
      return;
    }
    if (req.url === "/redirect-cookie") {
      const { port } = server.address();
      send(res, 302, undefined, {
        Location: `http://localhost:${port}/cookie`,
      });
      return;
    }
    if (req.url === "/loop") {
      loopHits += 1;
      send(res, 302, undefined, { Location: "/loop" });
      return;
    }
    if (req.url === "/loop-count") {
      // How many times /loop was asked for since the last read, so a test can
      // tell one request per hop from a transport quietly chaining hops.
      const body = Buffer.from(String(loopHits), "ascii");
      loopHits = 0;
      send(res, 200, body);
      return;
    }
    if (req.url === "/slow") {
      // Not answered within a test's patience: what a caller cancels against.
      const timer = setTimeout(
        () => send(res, 200, Buffer.from("slow response\n")), 30000);
      res.on("close", () => clearTimeout(timer));
      return;
    }

    let body;
    if (req.url === "/authorization") {
      body = Buffer.from(req.headers.authorization ?? "none", "ascii");
    } else if (req.url === "/cookie") {
      body = Buffer.from(req.headers.cookie ?? "none", "ascii");
    } else if (req.url === "/large") {
      body = largeBody;
    } else {
      body = Buffer.from("small response\n");
    }
    send(res, 200, body, {
      "Content-Type": "application/octet-stream",
      "X-Repeated": ["one", "two"],
    });
    return;
  }

  if (req.method === "POST" && req.url === "/early") {
    // Answer before consuming the upload. Draining the readable side after ending
    // the response lets the socket close with FIN instead of resetting and losing
    // the response bytes the client already parsed.
    req.on("error", () => {});
    req.resume();
    send(res, 413, undefined, { Connection: "close" });
    return;
  }

  if (req.method === "POST") {
    const chunks = [];
    req.on("data", (chunk) => chunks.push(chunk));
    req.on("end", () => send(res, 200, Buffer.concat(chunks)));
    return;
  }

  send(res, 405);
});

server.on("clientError", (_error, socket) => socket.destroy());
server.listen(0, "127.0.0.1", () => {
  const { port } = server.address();
  writeFileSync(portFile, String(port), "ascii");
});
